File: typescript/003_looping_chatbot.ts
// Load env Variables----------------------------------------------------
import Anthropic from "@anthropic-ai/sdk";
import * as dotenv from "dotenv";
import * as readline from "readline";

//================================================
// Looping Chatbot
//================================================

dotenv.config();


// Create Client-------------------------------------------------------
const client = new Anthropic();
const model = "claude-haiku-4-5";


// Helper Functions
const addUserMessage = (messages: Anthropic.MessageParam[], text: string) => {
    messages.push({ role: "user", content: text });
};

const addAssistantMessage = (messages: Anthropic.MessageParam[], text: string) => {
    messages.push({ role: "assistant", content: text });
};

// Makes Request-----------------------
const chat = async (messages: Anthropic.MessageParam[]): Promise<string> => {
    const message = await client.messages.create({
        model,
        max_tokens: 1000,
        messages,
    });
    return (message.content[0] as Anthropic.TextBlock).text;
};


const main = async () => {
    const messages: Anthropic.MessageParam[] = [];

    // readline replaces Python's input() for getting user input in Node.js
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const askQuestion = (prompt: string): Promise<string> =>
        new Promise((resolve) => rl.question(prompt, resolve));

    // Use a `while true` loop to run chatbot forever
    while (true) {
        // Get user input
        const userInput = await askQuestion("> ");
        console.log("> ", userInput);

        // Add user input to list of messages
        addUserMessage(messages, userInput);

        // Get Claude's response by calling claude with the chat function
        const answer = await chat(messages);

        // Add Claude's response to the list of messages
        addAssistantMessage(messages, answer);

        console.log("--------");
        console.log(answer);
    }
};

main();

File: python/003_looping_chatbot.py
#Load env Variables----------------------------------------------------
from dotenv import load_dotenv
from anthropic import Anthropic

#================================================
# Looping Chatbot
# 1. Prompt the user to enter some input
# 2. Add the user input to the list of messages
# 3. Call the API
# 4. Add generated text to the list of messages
# 5. Print the generated text
# 6. Repeat
#================================================

load_dotenv()


# Create Client-------------------------------------------------------
client = Anthropic()
model = "claude-haiku-4-5"


# Helper Functions
def add_user_message(messages, text):
    user_message = {"role": "user", "content": text}
    messages.append(user_message)

def add_assistant_message(messages, text):
    assistant_message = {"role": "assistant", "content": text}
    messages.append(assistant_message)

# Makes Request-----------------------
def chat(messages):
    message = client.messages.create(
        model=model,
        max_tokens=1000,
        messages=messages,
    )
    return message.content[0].text


messages = [] # message list

# Use a `while True` loop to run chatbot forever
while True:# 6. Repeat
    # 1. Prompt the user to enter some input
    user_input = input ("> ")
    print("> ", user_input)

    # 2. Add the user input to the list of messages
    add_user_message(messages, user_input)

    # 3. Call the API
    answer = chat(messages)

    # 4. Add generated text to the list of messages
    add_assistant_message(messages, answer)

    # 5. Print the generated text
    print("--------")
    print(answer)
